src/com/google/enterprise/adaptor/database/DatabaseAdaptor.java

```diff
@@ -723,6 +731,12 @@ static Acl buildAcl(ResultSet rs, String delim, String namespace)
         hasColumn(metadata, GsaSpecialColumns.GSA_PERMIT_GROUPS);
     boolean hasDenyGroups =
         hasColumn(metadata, GsaSpecialColumns.GSA_DENY_GROUPS);
+
+    if (!hasPermitUsers && !hasDenyUsers
+        && !hasPermitGroups && !hasDenyGroups) {
+      return defaultAcl;
```

Comment:
why do we need default acl parameter?

Reply:
If we have aclSql but no ACL entries, we want an empty ACL to mark the document secure. But if we have no aclSql and no ACL entries, then we want a null ACL to mark the document public.